pydantic/_internal/_model_construction.py

```diff
@@ -417,6 +425,12 @@ def inspect_namespace(  # noqa C901
             and ann_type not in all_ignored_types
             and getattr(ann_type, '__module__', None) != 'functools'
         ):
+            if isinstance(ann_type, str):
+                # Walking up the frames to get the module namespace where the model is defined
+                # (as the model class wasn't created yet, we unfortunately can't use `cls.__module__`):
+                module = inspect.getmodule(sys._getframe(2))
+                if module is not None:
+                    ann_type = eval_type_backport(ForwardRef(ann_type), globalns=module.__dict__)
```

Comment:
What if the model is defined in a function?

Reply:
It will fail indeed. Not having access to the current model class makes things annoying as stated in my comment. Not sure if we can address this issue? As I said in the original issue, the `Annotated` pattern isn't recommended anyway.

Reply:
I think we should consider many of the questions we've been thinking about for our other recent namespace management changes - how are things different if we rebuild or import a model in a new file?

Reply:
Should we be filtering the namespace with our new logic in `_typing_extra.py`?

Reply:
> how are things different if we rebuild or import a model in a new file?

In these cases we have access to the class' module, so it helps as we can easily combine the module namespace and parent namespace (if necessary).

I'll note that because the way `inspect.getmodule` works, the following is ok:

```python
from typing import Annotated

from pydantic import BaseModel, PrivateAttr

def func():
    class Model(BaseModel):
        _foo: 'Annotated[int, PrivateAttr()]'

func()
```

(however, it does not if `PrivateAttr` and/or `Annotated` is imported inside `func`).

---

We can consider many scenarios where even our current annotation evaluation for normal fields (i.e. after `cls` is created in the metaclass, where all the logic to get the parent ns and merge it with the module ns is done) does not work as expected:

```python
from __future__ import annotations

from pydantic import BaseModel


def test():
    from typing import Annotated

    def inner():

        from pydantic import Field

        class Model(BaseModel):
            a: Annotated[int, Field(default=1)]

        return Model

    Model = inner()

test()
```

Trying to support these cases would probably complicate things (and also maybe increase memory/cpu usage). I'm also wondering how much of these use cases can be covered by PEP 649.